src/App/ExtensionContainer.h

```diff
@@ -135,9 +135,72 @@ class AppExport ExtensionContainer : public App::PropertyContainer
     ExtensionT* getExtensionByType(bool no_except=false, bool derived=true) const {
         return static_cast<ExtensionT*>(getExtension(ExtensionT::getExtensionClassTypeId(),derived,no_except));
     }
+
+    /** Helpers for iterator over extensions of a given type */
+    //@{
+
+    /** Helper structs to deduce the class type of a class function pointer type */
+    //@{
+    template<typename T> struct ClassOf {};
+    template<typename R, typename C>
+        struct ClassOf<R (C::*)>{   using type = C;    };
+    //@}
+
+    /** Call an extension member function for each extension of the given type
+     * @param f: the member function pointer of some extension class
+     * @param args: variadic parameters passed to the member function
+     */
+    template<typename F, class... Args>
+    void callExtension(F&& f, Args&&... args) const {
+        for(auto entry : _extensions) {            
+            if(entry.first.isDerivedFrom(ClassOf<F>::type::getExtensionClassTypeId()))
+                (static_cast<typename ClassOf<F>::type*>(entry.second)->*f)(std::forward<Args>(args)...);
+        }
+    }
+
+    /** Query using an extension member function
+     * @param f: the member function pointer of some extension class
+     * @param args: variadic parameters passed to the member function
+     *
+     * @return Returns true if the iteratorion is terminated prematurely, or else false.
+     *
+     * The iteration over the contained extensions is terminated as soon as the
+     * extension method call returns true.
+     */
+    template<typename F, class... Args>
+    bool queryExtension(F&& f, Args&&... args) const {
+        for(auto entry : _extensions) {            
+            if(entry.first.isDerivedFrom(ClassOf<F>::type::getExtensionClassTypeId())) {
+                if((static_cast<typename ClassOf<F>::type*>(entry.second)->*f)(std::forward<Args>(args)...))
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    /** Invoke a callable for each extenions of a given type
+     * @param f: the callable that accepts one argument of an extension, and
+     * returns any value that is convertable to boolean.
+     *
+     * @return Returns true if the iteratorion is terminated prematurely, or else false.
+     *
+     * The iteration over the contained extensions is terminated as soon as the
+     * extension method call returns true.
+     */
+    template<typename T, typename F>
+    bool foreachExtension(F f) const {
+        for(auto entry : _extensions) {            
+            if(entry.first.isDerivedFrom(T::getExtensionClassTypeId())) {
+                if(f(static_cast<T*>(entry.second)))
+                    return true;
+            }
+        }
+        return false;
+    }

```

Comment:
I really like these helpers, great for readability of the other code. However, I would personally put these helpers, along with modifications to use them in existing code, in a separate commit, which would then make no functional changes, just optimize the existing code, making both resulting commits easier to review. It's of course not up to me and I believe that the form/separation of these PRs has already been discussed a lot, but just wanted to add my 2c here :-)

Reply:
+1